Add gRPC performance section on flow control (#25901)
dotnet/AspNetCore.Docs: aspnetcore/grpc/performance.md

:::moniker-end

## Flow control

HTTP/2 flow control is a feature that prevents apps from being overwhelmed with data. When using flow control:

* Each app has an available buffer window. The buffer window is how much data the app can receive at once.
* Flow control activates if the buffer window is filled up. When activated, the sending app pauses sending more data.

> * Once the receiving app has processed data then space in the buffer window is available. The sending app resumes sending data.

Reviewer: ```suggestion
* Once the receiving app has processed data, then space in the buffer window is available. The sending app resumes sending data.
```

> Flow control can have a negative impact on performance when sending and receiving large messages. If the buffer window isn't large enough, or there is latency between the client and server, then data can be sent in start/stop bursts. This can be fixed by increasing HTTP/2 flow control window size. In Kestrel this is configured with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialConnectionWindowSize> and <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialStreamWindowSize> at app startup.

Reviewer: ```suggestion
Flow control can have a negative impact on performance when sending and receiving large messages. If the buffer window isn't large enough or there's latency between the client and server, then data can be sent in start/stop bursts. This can be fixed by increasing the HTTP/2 flow control window size. In Kestrel, this is configured with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialConnectionWindowSize> and <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialStreamWindowSize> at app startup.
```

```csharp
builder.WebHost.ConfigureKestrel(options =>
{
    var http2 = options.Limits.Http2;
    http2.InitialConnectionWindowSize = 2 * 1024 * 1024 * 2; // 2 MB
    http2.InitialStreamWindowSize = 1024 * 1024; // 1 MB
});
```

> [!IMPORTANT]

> > Increasing Kestrel's window size allows the app to buffer more data and possibility increasing memory usage. Avoid configuring an unnecessarily large window size.

Reviewer: ```suggestion
> Increasing Kestrel's window size allows the app to buffer more data, which possibly increases memory usage. Avoid configuring an unnecessarily large window size.
```

:::moniker-end

## Flow control

HTTP/2 flow control is a feature that prevents apps from being overwhelmed with data. When using flow control:

> * Each app has an available buffer window. The buffer window is how much data the app can receive at once.

Reviewer 2: Each connection has a buffer window, and each stream also has one. The app does not.

Reviewer 3: I'm guessing "app" here means invocation of the middleware/procedure, but I think saying that the limit is for both the connection and request is clearer and more precise.

* Flow control activates if the buffer window is filled up. When activated, the sending app pauses sending more data.
* Once the receiving app has processed data, then space in the buffer window is available. The sending app resumes sending data.

> Flow control can have a negative impact on performance when sending and receiving large messages. If the buffer window is smaller than incoming message payloads, or there's latency between the client and server, then data can be sent in start/stop bursts.

Reviewer: The comma before the "or" is breaking the two objects of the conditional clause ("if"). 

```suggestion
Flow control can have a negative impact on performance when sending and receiving large messages. If the buffer window is smaller than incoming message payloads or there's latency between the client and server, then data can be sent in start/stop bursts.
```

> Flow control performance issues can be fixed by increasing buffer window size. In Kestrel, this is configured with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialConnectionWindowSize> and <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialStreamWindowSize> at app startup. Kestrel's default stream window size is 96 KB. If a gRPC service often receives messages larger than this then consider increasing the connection and stream window size.

Reviewer: ```suggestion
Flow control performance issues can be fixed by increasing buffer window size. In Kestrel, this is configured with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialConnectionWindowSize> and <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.InitialStreamWindowSize> at app startup. Kestrel's default stream window size is 96 KB. If a gRPC service often receives messages larger than 96 KB, consider increasing the connection and stream window size.
```

```csharp
builder.WebHost.ConfigureKestrel(options =>
{
    var http2 = options.Limits.Http2;
    http2.InitialConnectionWindowSize = 2 * 1024 * 1024 * 2; // 2 MB
    http2.InitialStreamWindowSize = 1024 * 1024; // 1 MB
});

> ```

Author: I don't want to go into quite that much detail, but I'll add something on the connection vs stream relationship.

> For more information about how flow control works, see [HTTP/2 Flow Control](https://medium.com/coderscorner/http-2-flow-control-77e54f7fd518) blog post.

Reviewer: ```suggestion
For more information about how flow control works, see [HTTP/2 Flow Control (blog post)](https://medium.com/coderscorner/http-2-flow-control-77e54f7fd518).
```

> [!IMPORTANT]

> > Increasing Kestrel's window size allows the app to buffer more data, which possibly increases memory usage. Avoid configuring an unnecessarily large window size.

Reviewer 3: ```diff
- > Increasing Kestrel's window size allows the app to buffer more data, which possibly increases memory usage. Avoid configuring an unnecessarily large window size.
+ > Increasing Kestrel's window size allows Kestrel to buffer more data on behalf of the app which possibly increases memory usage. Avoid configuring an unnecessarily large window size.
```
